### PR TITLE
Adding a Symmetric Crypto Service to ewasm

### PR DESCRIPTION
**Pull request by fredfortier**
- [x] Public interfaces
- [x] Resolver
- [x] Implementation in enclave

### CODE REVIEW

File: eng-wasm/src/crypto_wasm.rs
Comment (moriaab) [inline eng-wasm/src/crypto_wasm.rs]: It is better to use already existing constants in symmetric.rs

File: enigma-core/enclave/src/auto_ffi.rs
Comment (moriaab) [inline enigma-core/enclave/src/auto_ffi.rs]: Please remove this from the PR. The file's contents didn't change, but there wos a warning during the generation.

File: enigma-crypto/src/symmetric.rs
Comment (moriaab) [inline enigma-crypto/src/symmetric.rs]: What is the purpose of this constant in this file if it is not used in the code?

File: enigma-runtime-t/src/lib.rs
Comment (moriaab) [inline enigma-runtime-t/src/lib.rs]: Please implement the need `From` for errors. We did an effort to keep the runtime readable and free from conversions :) See https://github.com/enigmampc/enigma-core/blob/develop/enigma-tools-t/src/common/errors_t.rs for the reference.

File: enigma-runtime-t/src/lib.rs
Comment (moriaab) [inline enigma-runtime-t/src/lib.rs]: Please implement the need From for errors. We did an effort to keep the runtime readable and free from conversions :) See https://github.com/enigmampc/enigma-core/blob/develop/enigma-tools-t/src/common/errors_t.rs for the reference.

File: eng-wasm/src/crypto_wasm.rs
Comment (moriaab) [inline eng-wasm/src/crypto_wasm.rs:6]: Can these constants and the constants in symmetric.rs have the same single origin and not be copied? Duplication may cause big problems in the future, if only one of the places will be changed.

**Review by moriaab — COMMENTED** [review]
Great job!
Please add tests for the usage of this feature.

File: eng-wasm/src/crypto_wasm.rs
Comment (fredfortier) [reply eng-wasm/src/crypto_wasm.rs:6]: Yes, but where would you put them? `eng_wasm` does not seem to use any of the common modules.

File: enigma-crypto/src/symmetric.rs
Comment (fredfortier) [reply enigma-crypto/src/symmetric.rs]: It's used in `runtime_t`. I'm just trying to avoid re-defining constants that can be imported. Let me know if there's a better place to put shared constants.

**Comment by fredfortier** [comment]
@moriaab Where should I put the unit tests? I can't see to find any for the `eng_wasm` module.

File: enigma-crypto/src/symmetric.rs
Comment (fredfortier) [reply enigma-crypto/src/symmetric.rs]: Added directly in `runtime_t` for now. Please let me know what is the right place for this constant. I don't see any shared module between `eng_wasm` and `runtime_t`.

File: enigma-runtime-t/src/lib.rs
Comment (fredfortier) [reply enigma-runtime-t/src/lib.rs]: Done.

File: enigma-runtime-t/src/lib.rs
Comment (fredfortier) [reply enigma-runtime-t/src/lib.rs]: Done.

File: enigma-core/enclave/src/auto_ffi.rs
Comment (fredfortier) [reply enigma-core/enclave/src/auto_ffi.rs]: Done

File: eng-wasm/src/lib.rs
Comment (moriaab) [inline eng-wasm/src/lib.rs:133]: It is better to put the tests [here](https://github.com/enigmampc/enigma-core/blob/develop/enigma-core/app/src/wasm_u/wasm.rs#L61)

1. Create the contract that uses encrypt/decrypt services.
2. Put the contract in https://github.com/enigmampc/enigma-core/tree/develop/examples/eng_wasm_contracts
3. **In Cargo.toml of the contract use the local paths for `eng-wasm` and `eng-wasm-derive`** instead of crates from crates.io.
4. Create a test in wasm.rs. All the helper functions are there as well as examples of working with `ethabi::Token`. Use the existing tests for reference. Here is the simplest form of the test https://github.com/enigmampc/enigma-core/blob/develop/enigma-core/app/src/wasm_u/wasm.rs#L363.

Comment (fredfortier) [reply eng-wasm/src/lib.rs:133]: Done.

File: eng-wasm/src/crypto_wasm.rs
Comment (fredfortier) [inline eng-wasm/src/crypto_wasm.rs:23]: @moriaab I tried everything that I could to use a dynamic buffer but it only seems to work with a static buffer allocated with zeros. Please confirm that dynamic buffers are not supported. If true, I'll see if I can optimize by estimating the size of the encrypted data instead of using a constant.

File: enigma-core/app/src/wasm_u/wasm.rs
Comment (moriaab) [inline enigma-core/app/src/wasm_u/wasm.rs:392]: The meaning of test is not just invoke the functionality, but to compare the results with the expected ones.

Comment (fredfortier) [reply enigma-core/app/src/wasm_u/wasm.rs:392]: The purpose of this test is to verify encryption/decryption, not to transform a value and verify the output. The contract will panic if the value isn't decrypted correctly and the test will fail. 

Please understand that the reason why I didn't return an output is because it is redundant. You should not assume that I don't know what a test is. I simply used `test_print_simple` as a template as per your suggestion. If you prefer that each function invoked by `compile_deploy_execute` must always return an output, just tell me and I'll happily make that change.

Comment (moriaab) [reply enigma-core/app/src/wasm_u/wasm.rs:392]: Oh, you are right, sorry, my comment was wrong.

File: examples/eng_wasm_contracts/encryption/src/lib.rs
Comment (moriaab) [inline examples/eng_wasm_contracts/encryption/src/lib.rs:76]: I realized that `no_mangle` are meaningless in secret contracts, so I removed it from all our example contracts. Please remove `no_mangle` to avoid the confusion about this contract.
Please remove the comment part about `no_mangle`: lines 47,48.

File: examples/eng_wasm_contracts/encryption/src/lib.rs
Comment (moriaab) [inline examples/eng_wasm_contracts/encryption/src/lib.rs:56]: Can you please clarify what is the purpose of `pub_key`?

File: examples/eng_wasm_contracts/encryption/src/lib.rs
Comment (moriaab) [inline examples/eng_wasm_contracts/encryption/src/lib.rs:34]: The lines 27, 34 and 35 are commented out, if there is no potential use, maybe it is better to remove them.

File: eng-wasm/src/crypto_wasm.rs
Comment (moriaab) [reply eng-wasm/src/crypto_wasm.rs:6]: @fredfortier I think that maybe getting `SYMMETRIC_KEY_SIZE` and `IV` should be also part of runtime services? What do you think?

File: eng-wasm/src/crypto_wasm.rs
Comment (moriaab) [inline eng-wasm/src/crypto_wasm.rs:21]: @fredfortier I think that you can estimate the encrypted payload size. In block cypher it is equal to the plaintext message with completion of the last part to a block size.

File: eng-wasm/src/crypto_wasm.rs
Comment (moriaab) [inline eng-wasm/src/crypto_wasm.rs:20]: @fredfortier As far as I understand, it is impossible to use the dynamic buffers here. You should allocate a continuous region of WASM memory. This region is filled when the runtime function is called. The dynamic array cannot be managed from the runtime.

**Review by moriaab — APPROVED** [review]
I added more comments, will appreciate if you could address them before merging. 
Thank you.

File: examples/eng_wasm_contracts/encryption/src/lib.rs
Comment (fredfortier) [reply examples/eng_wasm_contracts/encryption/src/lib.rs:56]: It does not do anything with local encryption. It is just verifying that it is possible to obtain valid key material from the generated private key. I do use this method in the coinjoin contract to encrypt data externally without sharing the private key.

File: eng-wasm/src/crypto_wasm.rs
Comment (fredfortier) [reply eng-wasm/src/crypto_wasm.rs:21]: Are you ok if I keep the static size for now and create a new PR to implement this? I other things to finish before I can revisit this and I don't want to keep the PR open for too long.

Comment (moriaab) [reply eng-wasm/src/crypto_wasm.rs:21]: Sure.
To make your merge easy, I am waiting for your merge before I merge PR#204, which changes the structure of the runtime.

**Pull request merged**
